bsp/Src/I2C.c

```diff
@@ -0,0 +1,529 @@
+	{
+		__HAL_RCC_I2C3_CLK_DISABLE();
+
+		HAL_GPIO_DeInit(GPIOA, GPIO_PIN_8);
+		HAL_GPIO_DeInit(GPIOB, GPIO_PIN_4);
+
+		HAL_NVIC_DisableIRQ(I2C3_EV_IRQn); // Disable the I2C interrupt
+		HAL_NVIC_DisableIRQ(I2C3_ER_IRQn); // Diable the I2C interrupt
+	}
+	#endif
+}
+
+/**
+* @brief    Transmits data onto the I2C bus.
+* @param    hi2c           to structure that has config data
+* @param    deviceAddr     target device address
+* @param    pDataBuff      data buffer
+* @param    len            amount of data
+* @return   I2C_StatusTypeDef
+*/
+I2C_StatusTypeDef i2c_send(I2C_HandleTypeDef *hi2c,
+              uint8_t deviceAddr,
+              uint8_t* pDataBuff,
+              uint16_t len)
+{
+	// Confirm length is not larger than its designated payload size
+	if(hi2c->Instance == I2C1 && len > I2C1_MAX_PAYLOADS)
+	{
+		// ERROR
+		return I2C_BADSIZE;
+	}
+	#ifdef I2C2
+	else if(hi2c->Instance == I2C2 && len > I2C2_MAX_PAYLOADS)
+	{
+		// ERROR
+		return I2C_BADSIZE;
+	}
+	#endif
+	#ifdef I2C3
+	else if(hi2c->Instance == I2C3 && len > I2C3_MAX_PAYLOADS)
+	{
+		// ERROR
+		return I2C_BADSIZE;
+	}
+	#endif
+
+	// Update the latest I2C peripheral
+	update_last_i2c(hi2c);
+
+	if(HAL_I2C_GetState(hi2c) == HAL_I2C_STATE_READY)
+	{
+		return I2C_ConvertHALStatus(HAL_I2C_Master_Transmit_IT(hi2c, deviceAddr, pDataBuff, len));
+	}
+	
+	QueueHandle_t I2C_Queue = I2C1_Queue;
+	if(hi2c->Instance == I2C1) {}
+	#ifdef I2C2
+	else if (hi2c->Instance == I2C2)
+	{
+		I2C_Queue = I2C2_Queue;
+	}
+	#endif
+	#ifdef I2C3
+	else if (hi2c->Instance == I2C3)
+	{
+		I2C_Queue = I2C3_Queue;
+	}
+	#endif
+	else
+	{
+		// Handle invalid I2C peripheral
+		return I2C_BADINST;
+	}
+
+	if (len > uxQueueSpacesAvailable(I2C_Queue))
+	{
+		return I2C_QFULL;
+	}
+	
+	for(int packetsLeft = len - 1; packetsLeft >= 0; packetsLeft--)
+	{
+		DataInfo_t dataToEnqueue = {
+			.hi2c = hi2c,
+			.deviceAddr = deviceAddr,
+			.pDataBuff = pDataBuff[len - packetsLeft - 1],
+			.payloadsLeft = packetsLeft
+		};
+
+		if (xQueueSend(I2C_Queue, &dataToEnqueue, 0) != pdPASS)
+		{
+			// Handle queue full error
+			return I2C_QFULL;
+		}
+	}
+
+	return I2C_SENT;
+}
+
+/**
+* @brief    Receives data from the I2C bus.
+* @param    hi2c:            to structure that has config data
+* @param    deviceAddr      target device address
+* @param    pDataBuff      data buffer
+* @param    len            amount of data
+* @return   I2C_StatusTypeDef
+*/
+I2C_StatusTypeDef i2c_recv(I2C_HandleTypeDef *hi2c,
+              uint8_t deviceAddr,
+              uint8_t* pDataBuff,
+              uint16_t len)
+{
+	// Update the latest I2C peripheral
+	update_last_i2c(hi2c);
+
+    return I2C_ConvertHALStatus(HAL_I2C_Master_Receive_IT(hi2c, deviceAddr, pDataBuff, len));
+}
+
+/**
+ * @brief 	Internal function to send the value of the specified i2c
+ * peripheral's queue
+ *
+ * @param	i2c_peripheral Either I2C1, I2C1, or I2C3
+ */
+static I2C_StatusTypeDef queue_send(I2C_TypeDef *i2c_peripheral)

```

Comment:
queue_send is a weird name to use here because it occurs on a receive so it's kind of confusing. also, perhaps taking this stuff and putting it in the tx complete callback function is a good idea.

Reply:
queue_send means you are sending the last item on the corresponding queue. It happens on receive because you are forcing the next item to send when a receive is completed. If you still believe the name requires to be changed please respond here.

Reply:
https://www.freertos.org/Documentation/02-Kernel/04-API-references/06-Queues/03-xQueueSend

FreeRTOS uses "queue send" to refer to putting items inside of a queue. I was more referring to this being the opposite of what the function does (pop item from queue and send to i2c), so very prone to a misunderstanding. Something like just "send" or "transmit" might be better (especially since this is static so only accessible in this file).

CAN's version is:
`static void transmit(CAN_HandleTypeDef* handle)`

Reply:
changed to match CAN name of transmit